app/models/game.rb

```diff
@@ -87,6 +87,108 @@ def is_in_check?(player)
     return false
   end
 
+  def is_in_checkmate?(player)
+    return false if !is_in_check?(player)
+    if player == white_player
+      king = pieces.where(type: "King", player: white_player).first
+      return false if king.can_escape_from_check?
+
+      # Finds the piece that is currently threatening the king
+      king_x = king.position_x
+      king_y = king.position_y
+      threatening_piece = nil
+      black_player.pieces.where(game: self, is_active: true).each do |piece|
+        if piece.type != "King"
+          threatening_piece = piece if piece.is_valid_capture?(king_x, king_y) && piece.is_valid_move?(king_x, king_y)
```

Comment:
is_valid_capture? just checks if the piece is an enemy or not, it doesn't have any logic to see if the move is valid. so if i used just is_valid_capture? every piece would be a threatening piece.